### PR TITLE
test: run tests on python 3.12

### PR DESCRIPTION
**Pull request by mvadari**
## High Level Overview of Change

Title says it all.

### Context of Change

I was running something on Python 3.12 and wanted to make sure everything would work properly.

### Type of Change

- [x] Tests (You added tests for code that already exists, or your new feature included in this PR)

### Did you update CHANGELOG.md?

- [ ] Yes
- [x] No, this change does not impact library users

## Test Plan

CI passes.

### CODE REVIEW

File: .github/workflows/integration_test.yml [hunk 1]
```diff
@@ -6,7 +6,7 @@ env:
 
 on:
   push:
-    branches: [ master ]
+    branches: [master]
```
Comment (ckeshava) [inline .github/workflows/integration_test.yml:9]: shouldn't we use the `main` branch for integration testing?

Comment (mvadari) [reply .github/workflows/integration_test.yml]: Good catch - the branch was renamed but CI wasn't updated.

File: .github/workflows/unit_test.yml [hunk 2]
```diff
@@ -2,7 +2,7 @@ name: Unit test
 
 on:
   push:
-    branches: [master]
+    branches: [main]
   pull_request:
   workflow_dispatch:
```
Comment (ckeshava) [inline .github/workflows/unit_test.yml:8]: the `lint-and-type-check` job is executed only on Python 3.7. It shouldn't cause any issue unless there is a breaking change in the linter in between python versions. Should we change it? What do you think?

**Review by ckeshava — APPROVED** [review]

File: .github/workflows/unit_test.yml (hunk 2 above)
Comment (mvadari) [reply .github/workflows/unit_test.yml:8]: That is on purpose, to ensure backwards compatibility.

Comment (mvadari) [reply .github/workflows/unit_test.yml:8]: That change could also be done in a separate PR, since it's irrelevant to this PR.

**Review by jonathanlei — APPROVED** [review]

**Pull request merged**